csrank/metrics.py

```diff
@@ -123,3 +123,79 @@ def topk_acc(y_true, y_pred):
         return acc
 
     return topk_acc
+
+
+def dcg_gain(grading, max_grade):
+    """Maps a grading (from 0 to `max_grade`) to its DCG "gain".
+
+    The gain is defined as in the Discounted Cumulative Gain (DCG)
+    metric. This is always a value in [0, 1] that can be interpreted as
+    a probability.
+    """
+    return (2 ** grading - 1) / (2 ** max_grade)
+
+
+def err(gradings, probability_mapping):
```

Comment:
The other metrics follow the interface:
`def metric(y_true, y_pred):`
Where y_true is a (n_instances, n_objects) shaped array of integers.
Example: A row `[0, 3, 2, 1]` denotes a ranking where the ground truth ordering is A > D > C > B.

Is it possible to unify the metrics with respect to this interface?

Reply:
Well `err` internally really expects a list of the probabilities that each object is "the relevant one" or "the one the user would stop its search at". So in your example, we might work with

```
[0.8, 0, 0.05, 0.15]
```

Which would imply that `A > D > C > B`.


The way it is currently implemented we use "gradings" instead, so the "ground truth" could for your example could be

| Object | Grading |
| ------ | ------- |
| A      | 4       |
| B      | 1       |
| C      | 2       |
| D      | 3       |

Which would then be transformed to probabilities using the Discounted Cumulative gain:

| Object | Grading | Gain   |
| ------ | ------- | ------ |
| A      | 4       | 0.9375 |
| B      | 1       | 0.0625 |
| C      | 2       | 0.1875 |
| D      | 3       | 0.4375 |

And from here we can actually use `err`. Currently its possible to either pass a different mapping from gradings to probabilities. If we want to standardize on `y_pred` and `y_true`, we'd have to infer the gradings from `y_pred` and then assume some mapping (like dcg).

Should I do that?

I've also added some WIP module documentation that documents the `y_true` and `y_pred` convention.

Reply:
What about a closure, which constructs a metric based on a mapping provided by the user.
Similar to:
```python
def make_ndcg_at_k_loss(k=5):
    def ndcg(y_true, y_pred):
...
```

Reply:
I'm not sure what you mean here -- what is `ndcg` and what is `k`?

Reply:
I have adapted the tensorflow implementation to the metrics api. Instead of a closure-constructor, the functions can be supplied as parameters with suitable defaults. This way the average user can just use the defaults, while more advanced users can use `functools.partial` to generate different cascade metrics or use different mappings.

The plain-python implementation still needs adaptation, I've hold off on that for now until you've reviewed the tensorflow implementation.

Reply:
Similarly, the tests have to be adapted (although the doctests already cover some of the code as well).

Reply:
This implementation of ERR looks good to me.